source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
     }
 }
 
-//Fix return value allways good fo aligment
+//Fix return value to be always good for alignment
 static int heap_alloc_internal_check(int16_t alloc_size)
 {
-    int temp_int, temp_sector = 0, temp_size;
-    temp_size = sizeof(int);
+    int temp_sector = 0;
+
     if (heap_main == 0) {
         heap_failure(NS_DYN_MEM_HEAP_SECTOR_UNITIALIZED);
-        alloc_size = 0;
     } else if (alloc_size < 1) {
         heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-        alloc_size = 0;
-    }
-
-
-    if (alloc_size) {
-        //
-        temp_sector = alloc_size;
-        temp_int = (temp_sector % temp_size);
-        if (temp_int) {
-            temp_sector += (temp_size - temp_int);
-        }
-
-
-        temp_sector /= temp_size;
-
-        temp_int = (heap_size / temp_size);
-        temp_int -= 2;
-        if (temp_sector > temp_int) {
-            heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-            temp_sector = 0;
-        }
+    } else if (alloc_size > (heap_size - 2 * sizeof(int)) ) {
+        heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
+    } else if (alloc_size) {
+        temp_sector = (alloc_size + sizeof(int) - 1) / sizeof(int);
     }
     return temp_sector;
 }
 
+// Checks that sector length indicators equals
+// Sector has format: Length (word) | data area in words | Length (word)
+// For direction, use 1 for direction up and -1 for down
 static int8_t ns_sector_validate(int *sector_start, int direction)
 {
     int8_t ret_val = -1;
-    int size_start, size_end;
-    if (direction) {
-        //Up
-        size_start = *sector_start++;
-        if (size_start < 0) {
-            sector_start += -(size_start);
-        } else {
-            sector_start += size_start;
-        }
-        size_end = *sector_start;
+    int size_start = *sector_start;
+    if (direction > 0) {
+        sector_start += (1 + abs(size_start));
     } else {
-        size_end = *sector_start--;
-        if (size_end < 0) {
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
+        sector_start -= (1 + abs(size_start));
     }
-    if (size_start == size_end) {
+
+    if (size_start && size_start == *sector_start) {
```

Comment:
Ah, got it. Was confused because the test on size was done after using the size.

I'd be inclined to move the size == 0 rejection to the top. Particularly as you probably should also be more strict on the test there before trying to follow advance the pointer.

Two other corruption indications are abs(size_start) being > (heap_size / sizeof(int) -  2),  or size being INT_MIN, which would stop abs() working - check that first.

And when doing numeric comparisons, don't use boolean shorthands. It's clearer to write the comparison with zero, rather than taking advantage of zero having the special meaning "false" - most particularly don't mix implicit boolean tests with comparisons, as this "if" does.